Eager(er) leftShift in mimir

djspiewak opened the pull request:
What I did here is make all of the definedness calculation eager, which avoids the largest hotspot that we saw in the stack traces. I also optimized for the extremely common homogeneous shift case, which previously caused a quadratic expansion of delegation through `UnionRight`. Finally, `emitOnUndefined = true` now is ultimately handled by a very tight and mostly-intrinsified loop within `BitSet` for any shift with a high water mark less than 64 (and above 64 it's still pretty fast).

djspiewak commented:
[ch771]

alissapajer commented on blueeyes/src/main/java/quasar/precog/BitSet.java:
```diff
@@ -256,6 +298,40 @@ public void flip(int fromIndex, int toIndex) {
+     * invariant here is that, for all ranges of i such
+     * that (i % mod >= 0 && i % mod < mod), there is *at least*
+     * one value of i such that get(i) == true.
+     */
```
If we're flipping all the bits for which `i % mod == 0`, it's possible that we flip some of those bits from `1` to `0`. As such, I don't see how the invariant that `get(i) == true` holds for certain ranges?

Say we have `[1, 0, 0, 1, 0, 0, 1]` and we call `flipByMod(3)`. Do we not get a bitset of all `0`'s?

djspiewak replied:
`flip` is the wrong word. You're actually pointing out a bug that this function had previously. You'll notice that the actual implementation uses `|=` rather than `^=`.

alissapajer replied:
Once a `1` always a `1`.

alissapajer replied:
I'd argue that "flip" is a technical term when it comes to bits, so renaming this would be helpful. Maybe `setByMod`?

alissapajer replied:
I've actually changed my mind. This is kind of a flipper. Just not the flipper you might be expecting.

djspiewak replied:
Is it the flipper we deserve?

kRITZCREEK replied:
![](https://media.giphy.com/media/eo0IYeDLUTEfm/giphy.gif)

djspiewak replied:
I can add the examples. The condition though is much more precise than the examples, and it translates directly into the scalacheck conditional. If you prefer, I could add phrasing to describe the condition as "`i` projected into the mod ring".

rintcius replied:
Yeah, I meant the examples just as an illustration of the text.

>  If you prefer, I could add phrasing to describe the condition as "i projected into the mod ring"

👍

alissapajer commented on blueeyes/src/main/java/quasar/precog/BitSet.java:
```diff
@@ -256,6 +298,40 @@ public void flip(int fromIndex, int toIndex) {
+                bits[i] = replacement;
+            }
+            return;
+        }
```
An alternate way to implement this would be to call `setBits` and provide an array of length `_length` filled entirely with `replacement`. That would actually require allocating a second array though, whereas what you have here is edit-in-place. So :+1:

djspiewak replied:
Yeah I thought about that approach. I *think* the way I'm doing it is faster, especially since it's extremely transparent to both HotSpot and the CPU itself, but I don't have measurements to back up that assumption.

alissapajer commented on blueeyes/src/main/java/quasar/precog/BitSet.java:
```diff
@@ -256,6 +298,40 @@ public void flip(int fromIndex, int toIndex) {
+            do {
+                if ((bits[i] & mask) == 0L) {
+                    bits[i] |= flipper;
+                }
```
I have parsed this to be:

`mask` is all `1`'s for an interval of length `mod`. We thus enter this `if` body when _none_ of the bits of `bits[i]` in that interval are `1`, in which case we set the first bit in that interval to `1` with the `flipper`. If we didn't enter the `if` body, we know at least one bit in that interval is already `1`, but it might not be the first one.

Is this what you're going for? Furthermore, the `0`th long array in `bits` is masked by a mask starting with the masking interval at the low-order. Then the `1`th long array in `bits` is masked with the interval rotated over once toward the high-order by `mod`, etc.

... this is nifty to think about. How does it fit in the larger picture of `Column`? :smile:

djspiewak replied:
This is indeed what I was going for! However, I just realized that there's a bug in it! I wonder if this is what's creating the failing tests. `mask` rotates around and will end up sometimes masking high-order *and* low-order bits simultaneously. This is in and of itself correct, but the high-order bits from `bits[i]` should be masked while the low-order bits from `bits[i + 1]` are masked, whereas right now everything is from `bits[i]` until it fully wraps around. This could be creating the bugs that we're seeing.

> How does it fit in the larger picture of Column? 😄

Great question! Think about the columns in a `LeftShift` hole which are not included in the `struct`. These columns need to be inflated such that each row in the input corresponds to at most *n* rows in the output, where *n* is the number of columns described by the struct (e.g. the width of the array/object). Note that the dataset may be heterogeneous, so *n* may not be the right size for *every* row, but it's the definitive high water mark.

Now, when we are doing `OnUndefined.Omit`, we can take the definedness bitset from the results of the shift and use that to mask the inflated unfocused columns (the ones which expand to size *n*). This gives us "inner" semantics on the map which is passed into the `repair` and strips out rows for which the shift was undefined. This is actually really easy and efficient.

When we do `Emit` though, we have to figure out *which* of the rows to hang onto. If `RightSide` is defined for a given *input* row, then we should throw away all of the others. If `RightSide` is undefined, then we want to keep *exactly* one row from the `LeftSide`. This is where it gets a little tricky, and it's why I wrote the soon-to-be-renamed `setByMod`.

When we're in the `Emit` case, we take the definedness bitset from the `RightSide` and call `setByMod` on it, where the `mod` is the high water mark (*n* from earlier). So this ensures that, for all input rows, *at least* one row will be defined in the unfocused columns. If a row is already defined (e.g. because `RightSide` is defined for that row), then we use it. Otherwise, we arbitrarily pick the first row and bring it back to life (it's valid to pick any of the rows within the mod ring, the first is just convenient).

rintcius commented on yggdrasil/src/test/scala/quasar/yggdrasil/LeftShiftSpec.scala:
```diff
@@ -199,6 +199,33 @@ trait LeftShiftSpec extends TableModuleTestSupport with SpecificationLike {
+      fromSample(
+        SampleData(
+          Stream(
+            toRecord(Array(0), JObject("b" -> JArray(JNum(1) :: JNum(2) :: Nil), "c" -> JArray(JNum(11) :: JNum(12) :: Nil)).some),
```
could improve the test a bit by making each record have a different index, i.e. next line becomes:

```
toRecord(Array(1), ...
```

djspiewak commented:
Mostly for my own notes… Here's a minimized version of the failing test: https://gist.github.com/26fa058cf0a5a390fc4981067329d62a

djspiewak commented on yggdrasil/src/main/scala/quasar/yggdrasil/table/ColumnarTableModule.scala:
```diff
@@ -1402,49 +1400,56 @@ trait ColumnarTableModule
+          innerIndex: Map[CPathNode, Int],
+          expansion: CF1,
+          highWaterMark: Int)
           : Map[ColumnRef, Column] = {
```
It is not. The standard is the way that I formatted it.

rintcius replied:
ok, written to memory :) there are places that don't follow this format, but the scalafmt PR did indeed format it like that.

djspiewak commented on blueeyes/src/test/scala/quasar/precog/BitSetSpec.scala:
```diff
@@ -0,0 +1,111 @@
+          (0 until (bound / mod) by mod) forall { i =>
+            (0 until mod) must contain { j: Int =>
+              if (i + j < bound)    // mod might not divide bound
+                bs(i + j) must beTrue
```
For the scalacheck property, I was only checking the guaranteed invariant 1. Checking invariant 2 is probably a good idea as well, but it's much harder to quantify. I check this in the other tests.

rintcius replied:
i actually thought it'd be _easier_ to verify. Just need to iterate over the bits:

- if it's possibly changed (i.e. the first bit in each range), then skip testing it
- for all other bits in these ranges they must be equal to the original bit

djspiewak commented on blueeyes/src/test/scala/quasar/precog/BitSetSpec.scala:
```diff
@@ -0,0 +1,111 @@
+          val bound = bs.length << 6
+
+          (0 until bound) forall { i =>
+            bs(i) mustEqual sparsened(i * mod + offset)
```
This property checks the only *guarantee* of the function.

rintcius replied:
hmm, why isn't there a guarantee that the bits of `sparsened` that are not checked in l.47 are not set?

djspiewak commented on blueeyes/src/test/scala/quasar/precog/BitSetSpec.scala:
```diff
@@ -0,0 +1,111 @@
+      (mod0 > Int.MinValue && offset0 > Int.MinValue) ==> {
+        val mod = math.abs(mod0) % 20    // don't let scalacheck go crazy
+
+        mod > 2 ==> {
```
I think I could change it to `mod > 1`. Allowing `mod = 1` would be weird and doesn't actually line up with what we care about.

djspiewak commented on blueeyes/src/test/scala/quasar/precog/BitSetSpec.scala:
```diff
@@ -0,0 +1,111 @@
+
+    "setByMod" in prop { (bs: BitSet, mod0: Int) =>
+      (mod0 > Int.MinValue) ==> {
+        val mod = math.abs(mod0) % 20    // don't let scalacheck go crazy
```
I'm probably going to push up this value, yes. The other tests are hitting a lot of the bizarre cases.

djspiewak commented on blueeyes/src/main/java/quasar/precog/BitSet.java:
```diff
@@ -217,6 +217,48 @@ public BitSet copy() {
+     * for all values of bs, i, o, m.
+     */
+    public BitSet sparsenByMod(final int offset, final int mod) {
+        long[] _bits = new long[_length * mod];
```
I could, but there's actually no point since we're never going to give it those values and anything `mod < 0` will lead to an exception anyway. I was actually thinking about removing the checks from `setByMod` as well for the same reason. This is not intended to be safe code.

rintcius replied:
hmm, I think it's good to have the checks, so that we are explicit in how it can be called & fail fast if someone calls it in a wrong way.

djspiewak replied:
Some of this is performance-sensitive code though. In this case, I think it might not matter much, but it's something to keep in mind. Checks aren't free.

djspiewak commented on blueeyes/src/main/java/quasar/precog/BitSet.java:
```diff
@@ -256,6 +298,83 @@ public void flip(int fromIndex, int toIndex) {
+     */
+    public void setByMod(int mod) {
+        if (mod >= 64) {
+            throw new IllegalArgumentException("this algorithm doesn't work for mod > 64; got " + mod);
```
Yeah the conditional is wrong, while the message is correct. Doesn't matter because the whole conditional is going away.

djspiewak commented:
[ch769]

This one should *also* be fixed by this change.

djspiewak commented:
[ch731]

djspiewak commented:
[ch746]

alissapajer commented:
@djspiewak before closing stories that are specific to particular queries failing, I think we should make sure there is a test hitting that query.

djspiewak commented:
@alissapajer Both of those only hard-fail due to externalities (timeouts and paging, respectively). I gave it some thought but I'm not sure how to write a test for them within our existing harnesses. Both could be integration tested using a hypothetical harness that sits on top the rest API in sdbe, but we don't have that harness right now. So I hate to say it, but we have to leave those to QA. :-(

djspiewak commented:
@rintcius I think I've addressed all your PR review feedback. Due to the pressing release schedule, I'm going to reassign this to someone in a western timezone so we can merge it tonight. Please feel free to leave more review feedback at your discretion and I'll address asap in a follow-up PR.

wemrysi commented on blueeyes/src/test/scala/quasar/precog/BitSetSpec.scala:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    "setByMod works with mod 3" in {
+      val bs = new BitSet(128)    // ensure we have seven longs
```
s/seven/two/

djspiewak replied:
They're basically the same number…

wemrysi commented on blueeyes/src/test/scala/quasar/precog/BitSetSpec.scala:
```diff
@@ -0,0 +1,153 @@
+    "setByMod works with mod 3" in {
+      val bs = new BitSet(128)    // ensure we have seven longs
+
+      // 64 % 35 = 29, and 64 * 6 = 384, and 384 % 35 = 0
```
Copy-paste artifact?

djspiewak replied:
It is.

rintcius commented:
@djspiewak awesome! The main comment that wasn't addressed was regarding the spec of `sparsenByMod`. I thought to just PR it to both verify my understanding of that function and explain what I mean - see #3738 

While fiddling with that spec I found a bug in `sparsenByMod`, see that PR for details.
